server/monitor-types/tls.js

```diff
@@ -0,0 +1,266 @@
+const { MonitorType } = require("./monitor-type");
+const { log, UP } = require("../../src/util");
+const net = require("net");
+const tls = require("tls");
+
+class TlsMonitorType extends MonitorType {
+
+    name = "port-tls";
+
+    /**
+     * Performs the periodic monitoring check on a TLS TCP port.
+     * @param {Monitor}          monitor   Monitor to check
+     * @param {Heartbeat}        heartbeat Monitor heartbeat to update
+     * @param {UptimeKumaServer} _server   Uptime Kuma server (unused)
+     * @returns {Promise<void>} A fulfilled promise if the check succeeds, a rejected one otherwise
+     */
+    async check(monitor, heartbeat, _server) {
+        const options = {
+            hostname: monitor.hostname,
+            port: monitor.port,
+            useStartTls: monitor.startTls || false,
+            request: monitor.request || null,
+            interval: monitor.interval || 30,
+        };
+
+        const abortController = new AbortController();
+
+        const timeoutMs = options.interval * 1000 * 0.8;
+        const timeoutID = setTimeout(() => {
+            this.log(`timeout after ${timeoutMs} ms`);
+            abortController.abort();
+        }, timeoutMs);
+
+        const tlsSocket = await this.connect(abortController.signal, options.hostname, options.port, options.useStartTls);
+        let tlsSocketClosed = false;
+        tlsSocket.on("close", () => {
+            tlsSocketClosed = true;
+        });
+
+        const result = await this.getResponseFromTlsPort(abortController.signal, tlsSocket, options)
+            .then((response) => {
+                clearTimeout(timeoutID);
+                return response;
+            })
+            .catch((error) => {
+                clearTimeout(timeoutID);
+                throw error;
+            })
+        ;
+
+        if (!tlsSocketClosed) {
+            tlsSocket.end();
+            tlsSocket.destroy();
+            this.debug_log("TLS socket commanded to close");
+        }
+
+        this.processResponse(result, monitor, heartbeat);
+    }
+
+    /**
+     * Compares the server's response against the monitor's attributes, decides on success/failure,
+     * and updates the heartbeat accordingly.
+     * @param {string}    response  Response received from the server
+     * @param {Monitor}   monitor   Monitor to check
+     * @param {Heartbeat} heartbeat Monitor heartbeat to update
+     * @returns {void}
+     * @throws Error if the check fails
+     */
+    processResponse(response, monitor, heartbeat) {
+        if (response instanceof Error) {
+            this.log("check failed: " + response.message);
+            throw response;
+        } else {
+            let success = false;
+            let message = undefined;
+            if (monitor.keyword) {
```

Comment:
nit: I think this part is a bit confusing to follow.
Please inline the logic from below

Reply:
I'm afraid inlining doesn't make it much clearer – to my eyes anyway:

```js
        const keywordContained = monitor.keyword && response.includes(monitor.keyword);
        const success = !monitor.keyword || (keywordContained === !monitor.invertKeyword);
        const message = monitor.keyword
                      ? keywordContained ? "Data contains keyword" : "Data does not contain keyword"
                      : "Connection successful";
```

So I broke it out into a separate function:

```js
    /**
     * Checks if the server response should be considered a success or not.
     * @param {string}  response Server response string
     * @param {Monitor} monitor  Monitor to check
     * @returns {[boolean, string]} Tuple with a boolean success indicator and a message string
     */
    checkResponseSuccess(response, monitor) {
        if (monitor.keyword) {
            // Monitor keyword present => Check if keyword is present/absent in response, depending
            // on whether the 'Invert Keyword' option is enabled.
            const keywordContained = response.includes(monitor.keyword);
            const message = keywordContained ? "Data contains keyword" : "Data does not contain keyword";
            return [ keywordContained === !monitor.invertKeyword, message ];
        } else {
            // No monitor keyword => Any response is considered a success.
            return [ true, "Connection successful" ];
        }
    }
```

This leaves `processResponse` like this:

```js
    processResponse(response, monitor, heartbeat) {
        if (response instanceof Error) {
            this.log("check failed: " + response.message);
            throw response;
        }

        const [ success, message ] = this.checkResponseSuccess(response, monitor);
        if (success) {
            this.log("check successful: " + message);
            heartbeat.msg = message;
            heartbeat.status = UP;
        } else {
            this.log("check failed: " + message);
            throw new Error(message);
        }
    }
```

What do you think?

Reply:
I was thinking of something like the following.
No need to extract or to make this less readable. What do you think?
(note that the comment was a `nit` => you can ignore it if you think that is better)

```js
    processResponse(response, monitor, heartbeat) {
        if (response instanceof Error) {
            log.info("tcp-keyword", "check failed: " + response.message);
            throw response;
        }
        if (!monitor.keyword) {
            heartbeat.msg = "Connection successful";
            heartbeat.status = UP;
            return;
        }
        const keywordContained = response.includes(monitor.keyword);
        if (keywordContained !== monitor.invertKeyword) {
            log.debug("tcp-keyword", "check successful: Data contains keyword");
            heartbeat.msg = "Data contains keyword";
            heartbeat.status = UP;
        } else {
            log.info("tcp-keyword", `check failed: keyword not contained in data: '${response.substring(0,30)}...'`);
            throw new Error(`Keyword not contained in data: '${response.substring(0,30)}...'`);
        }
    }

Reply:
Yeah, this isn't bad either. My strategy was to clearly separate the "success check" from the handling of the heartbeat. That way, if the heartbeat API ever changes, there's only a single location to change. But I'm definitely going to revamp the messages a bit to be more informational. :+1: